### PR TITLE
[TS migration] Migrate 'Reactions' component to TypeScript

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

[TS migration] Migrate 'Reactions' component to TypeScript

### Fixed Issues

$ https://github.com/Expensify/App/issues/25153
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

1. Open the app using account A.
2. Go to chat with user B. Send any message.
3. WEB/DESKTOP: Hover over the message, the context menu should appear in the top right corner of the message. 
3.1 Make sure it looks the same way as before and has quick reaction options in the list. 
3.2 When you hover over the reaction you should see a small tooltip with reaction info.
3.3 Click on any reaction button in the menu , "👍"  for example. The "👍" gets added as a reaction to the message.
When pressing the "👍" in the menu again, the reaction should be removed.
4. Right press (on web/desktop), or long-press (mobile) the message -> the context menu should be opened. It should have several quick reactions and _Add Reactions_ button at the top.
4.1 Add several reactions using quick reactions and _Add Reactions_ button.
5. Make sure reactions to the message look ok.
6. Login from account B on another device. Open the same chat and make sure reactions look the same way. Add a reaction to the same message, make sure it looks ok.
7. Go back to using account B.
8. WEB/DESKTOP: Hover over the reaction bubble of the message, you should see a tooltip with the info about users who left the reactions, make sure it looks the same way as before.
9. MOBILE: Long press the reaction bubble of the message, you should see a popup with the info about users who left the reactions, make sure it looks the same way as before.
10. Press _Add Reactions_  button in the bottom of the message -> Change default skin tone -> Select another skin tone.
11. Use reactions with another skin tone, make sure it works ok.

### Offline tests

N/A

### QA Steps

- [x] Verify that no errors appear in the JS console

Same as in the **Tests** section.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/23176449/6a862965-fb97-4f18-b93a-02437c5d4bc1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/23176449/9598fefd-70ac-4071-88bf-2fe0daf9f211

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/23176449/79e3af5a-cf27-45df-a9e2-788b89f99a05

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/23176449/b1f87560-c446-4574-b4b4-7d8e05125286

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/23176449/3f85cdcd-dce0-4818-ab9c-1ed3e5406479

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/23176449/aa87ab12-5bd5-4a0b-87e1-a9b96f6cdbad

</details>

### CODE REVIEW

File: src/components/Reactions/QuickEmojiReactions/types.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import {TextInput} from 'react-native';
+import type {OnyxEntry} from 'react-native-onyx';
+import type {Emoji} from '@assets/emojis/types';
+import type {AnchorOrigin} from '@userActions/EmojiPickerAction';
+import type {Locale, ReportAction, ReportActionReactions} from '@src/types/onyx';
+
+// TODO: remove when ReportActionContextMenu file migration https://github.com/Expensify/App/pull/32670 is merged
```
Comment (fabioh8010) [inline src/components/Reactions/QuickEmojiReactions/types.ts:7]:
```suggestion
// TODO: Remove this once AnchorForCommentsOnly (https://github.com/Expensify/App/issues/25135) is migrated to TypeScript.
```

To be better aligned with the docs.

**Review by fabioh8010 — APPROVED** [review]

File: src/components/Reactions/MiniQuickEmojiReactions.tsx [hunk 2]
```diff
@@ -1,106 +1,94 @@
...
 import * as EmojiPickerAction from '@userActions/EmojiPickerAction';
 import * as Session from '@userActions/Session';
 import CONST from '@src/CONST';
 import ONYXKEYS from '@src/ONYXKEYS';
-import {baseQuickEmojiReactionsDefaultProps, baseQuickEmojiReactionsPropTypes} from './QuickEmojiReactions/BaseQuickEmojiReactions';
+import type {BaseQuickEmojiReactionsOnyxProps, BaseQuickEmojiReactionsProps} from './QuickEmojiReactions/types';
 
-const propTypes = {
-    ...baseQuickEmojiReactionsPropTypes,
+type MiniQuickEmojiReactionsOnyxProps = Omit<BaseQuickEmojiReactionsOnyxProps, 'preferredLocale'>;
```
Comment (blazejkustra) [inline src/components/Reactions/MiniQuickEmojiReactions.tsx:20]: Why `preferredLocale` is omitted? I see that it is used in the component

File: src/components/Reactions/QuickEmojiReactions/index.native.tsx [hunk 3]
```diff
@@ -1,29 +1,18 @@
...
-     * Function that can be called to close the
-     * context menu in which this component is
-     * rendered.
-     */
-    closeContextMenu: PropTypes.func.isRequired,
-};
-
-function QuickEmojiReactions(props) {
-    const onPressOpenPicker = (openPicker) => {
+function QuickEmojiReactions({closeContextMenu, ...props}: QuickEmojiReactionsProps) {
```
Comment (blazejkustra) [inline src/components/Reactions/QuickEmojiReactions/index.native.tsx:6]: NAB:
```suggestion
function QuickEmojiReactions({closeContextMenu, ...rest}: QuickEmojiReactionsProps) {
```

File: src/components/Reactions/QuickEmojiReactions/index.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import {contextMenuRef} from '@pages/home/report/ContextMenu/ReportActionContextMenu';
+import CONST from '@src/CONST';
+import BaseQuickEmojiReactions from './BaseQuickEmojiReactions';
+import type {OpenPickerCallback, QuickEmojiReactionsProps} from './types';
+
+function QuickEmojiReactions({closeContextMenu, ...props}: QuickEmojiReactionsProps) {
```
Comment (blazejkustra) [inline src/components/Reactions/QuickEmojiReactions/index.tsx:7]: NAB:
```suggestion
function QuickEmojiReactions({closeContextMenu, ...rest}: QuickEmojiReactionsProps) {
```

File: assets/emojis/index.ts [hunk 5]
```diff
@@ -26,10 +29,10 @@ const emojiCodeTableWithSkinTones = emojis.reduce<EmojiTable>((prev, cur) => {
     return newValue;
 }, {});
 
-const localeEmojis = {
+const localeEmojis: LocaleEmojis = {
```
Comment (blazejkustra) [inline assets/emojis/index.ts:32]: Why was this needed?

**Review by blazejkustra — COMMENTED** [review]
LGTM, just couple questions

File: src/components/Reactions/MiniQuickEmojiReactions.tsx (hunk 2 above)
Comment (VickyStash) [reply src/components/Reactions/MiniQuickEmojiReactions.tsx:20]: Yeah, but it's omitted from OnyxProps type, cause this component doesn't take `preferredLocale` from `withOnyx` HOC.
If it looks confusing I can update this type to be just
```
type MiniQuickEmojiReactionsOnyxProps = {
    /** All the emoji reactions for the report action. */
    emojiReactions: OnyxEntry<ReportActionReactions>;

    /** The user's preferred skin tone. */
    preferredSkinTone: OnyxEntry<string | number>;
};
```

Comment (blazejkustra) [reply src/components/Reactions/MiniQuickEmojiReactions.tsx:20]: Yes, let's use a separate type 👍

File: assets/emojis/index.ts (hunk 5 above)
Comment (VickyStash) [reply assets/emojis/index.ts:32]: Otherwise, I had a TS error in [this function](https://github.com/VickyStash/Expensify/blob/ts-migration/reactions-component/src/libs/EmojiUtils.ts#L57). I guess it also could be resolved with assertion/additional conditions if you think type update isn't a good idea

Comment (blazejkustra) [reply assets/emojis/index.ts:32]: It's fine as you coded 👍

**Review by blazejkustra — APPROVED** [review]

**Review by fedirjh — APPROVED** [review]
Looking good and tests well.

**Review by AndrewGable — APPROVED** [review]

File: src/types/onyx/ReportActionReactions.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ type ReportActionReaction = {
     users: UsersReactions;
 
     /** Is this action pending? */
-    pendingAction?: OnyxCommon.PendingAction;
+    pendingAction: OnyxCommon.PendingAction;
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/types/onyx/ReportActionReactions.ts:27]: @fedirjh Right, I've prepared a [follow-up PR](https://github.com/Expensify/App/pull/34095)
